### DISCUSSION TITLE
iOS App Store requirements

### DISCUSSION THREAD

**Discussion by ehenighan** [Instawork/hyperview#696] [Q&A] [answered]
Hi,

Just wondered if there was any specific documentation/confirmation around section 2.5.2 of the Apple App Store review guidelines (https://developer.apple.com/app-store/review/guidelines/#software-requirements)

`2.5.2 Apps should be self-contained in their bundles, and may not read or write data outside the designated container area, nor may they download, install, or execute code which introduces or changes features or functionality of the app, including other apps. Educational apps designed to teach, develop, or allow students to test executable code may, in limited circumstances, download code provided that such code is not used for other purposes. Such apps must make the source code provided by the app completely viewable and editable by the user.`

I personally read this as inapplicable, because all the code executed _is_ built in and is just _executed_ based on markup from the server - however, there's just enough ambiguity around it e.g. with respect to "...which introduces or changes features or functionality of the app" that I'd like to ask the question! It seems like it must be fine given that e.g. AirBnB are doing something similar with their own custom framework. However, they might be big enough to get special dispensation, etc. In particular, is there any guidance about how you can/should update the app description/'release' notes to account for new functionality, when there's no new app version itself?

Sorry if this has been asked and answered before - I just can't really find much good information on the subject generally.

Kind regards,

Ed

**Comment by ehenighan** [comment]
Just following up in a bit more detail having gone back through the developer agreement:

https://developer.apple.com/support/terms/apple-developer-program-license-agreement/#apis-functionality

I read 3.3.2 as supportive because React Native -> JavaScript -> interpreted runtime, and you'd realistically only ever want to change your functionality in ways that were consistent with your main purpose! However, 3.3.3 is a bit more ambiguous:

`3.3.3 Without Apple’s prior written approval or as permitted under Section 3.3.25 (In-App Purchase API), an Application may not provide, unlock or enable additional features or functionality through distribution mechanisms other than the App Store, Custom App Distribution or TestFlight.`

Is server-side rendering a 'distribution mechanism' for the purposes of this clause? I doubt it when read in context of the previous clause, but...

**Reply by mdwagner** [reply, marked as answer]
I interpret it as: SSR via XML isn't actually providing, unlocking, or enabling additional features or functionality, because all of the functionality is already included in the distribution that's shipped to the App Store. It's no different than JSON including some data that is used in a conditional in the app.

The best example I could think of is if you wanted to add support for the Camera API, you would have to release a new App Store version, because the XML doesn't contain any functionality. But if you decide to remove some text in a description, your API just sends different text.

Release notes are really only needed for the included functionality, not necessarily for adding content.

But I understand where you're coming from, and to be fair, not many apps are even built this way.

**Reply by adamstep** [reply]
This is our understanding as well: the XML responses are not modifying the binary code to unlock new iOS SDKs. To access additional iOS features, you would need to release a new app version with updated app code. I wrote more about this previously here: https://github.com/Instawork/hyperview/issues/247
